File: src/exporter.py
"""
exporter.py — 추적 데이터를 JSON + After Effects Keyframe Data 로 저장
"""

import json
import os
from typing import List, Callable
try:
    from .tracker import FrameData, VideoInfo, Point2D, PersonData, HAND_LANDMARK_NAMES, POSE_LANDMARK_NAMES
except ImportError:
    from tracker import FrameData, VideoInfo, Point2D, PersonData, HAND_LANDMARK_NAMES, POSE_LANDMARK_NAMES


# ──────────────────────────────────────────────────────────────────────────
# JSON 내보내기
# ──────────────────────────────────────────────────────────────────────────

def export_json(frames: List[FrameData], info: VideoInfo, out_path: str,
                include_face: bool = True, include_body: bool = True,
                include_hands: bool = True) -> bool:
    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)

    def pt(p: Point2D) -> dict:
        return {"x": round(p.x, 3), "y": round(p.y, 3), "confidence": round(p.confidence, 4)}

    data = {
        "metadata": {
            "fps":          info.fps,
            "total_frames": info.total_frames,
            "width":        info.width,
            "height":       info.height,
            "included":     {"face": include_face, "body": include_body, "hands": include_hands},
        },
        "landmark_info": {
            "face": {
                "total_points": 478,
                "named": {
                    "right_eye_outer":  33,
                    "right_eye_inner":  133,
                    "right_iris":       473,
                    "left_eye_inner":   362,
                    "left_eye_outer":   263,
                    "left_iris":        468,
                    "nose_bridge_top":  168,
                    "nose_tip":         4,
                    "mouth_right":      61,
                    "mouth_upper":      13,
                    "mouth_left":       291,
                    "mouth_lower":      14,
                }
            },
            "hand": {
                "total_points": 21,
                "names": HAND_LANDMARK_NAMES,
            },
            "pose": {
                "total_points": 33,
                "names": POSE_LANDMARK_NAMES,
                "key_joints": {
                    "left_shoulder": 11, "right_shoulder": 12,
                    "left_elbow":    13, "right_elbow":    14,
                    "left_wrist":    15, "right_wrist":    16,
                    "left_hip":      23, "right_hip":      24,
                    "left_knee":     25, "right_knee":      26,
                    "left_ankle":    27, "right_ankle":     28,
                }
            },
        },
        "frames": []
    }

    def hand_to_dict(h):
        d = {"detected": h.detected, "side": h.side}
        if h.detected:
            d["landmarks"] = [
                {**pt(lm), "index": i, "name": HAND_LANDMARK_NAMES[i]}
                for i, lm in enumerate(h.landmarks)
            ]
        return d

    for fd in frames:
        jf = {"frame": fd.index, "timestamp": round(fd.timestamp, 4), "persons": []}

        for person in fd.persons:
            jp = {"person_id": person.person_id}

            if include_face:
                jface = {"detected": person.face.detected}
                if person.face.detected:
                    jface["named"] = {
                        "right_eye_outer":  pt(person.face.right_eye_outer),
                        "right_eye_inner":  pt(person.face.right_eye_inner),
                        "right_iris":       pt(person.face.right_iris),
                        "left_eye_inner":   pt(person.face.left_eye_inner),
                        "left_eye_outer":   pt(person.face.left_eye_outer),
                        "left_iris":        pt(person.face.left_iris),
                        "nose_bridge_top":  pt(person.face.nose_bridge_top),
                        "nose_tip":         pt(person.face.nose_tip),
                        "mouth_right":      pt(person.face.mouth_right),
                        "mouth_upper":      pt(person.face.mouth_upper),
                        "mouth_left":       pt(person.face.mouth_left),
                        "mouth_lower":      pt(person.face.mouth_lower),
                    }
                    jface["all_landmarks"] = [pt(p) for p in person.face.all]
                jp["face"] = jface

            if include_hands:
                jp["left_hand"]  = hand_to_dict(person.left_hand)
                jp["right_hand"] = hand_to_dict(person.right_hand)

            if include_body:
                jpose = {"detected": person.pose.detected}
                if person.pose.detected:
                    jpose["key_joints"] = {
                        "left_shoulder":  pt(person.pose.left_shoulder),
                        "right_shoulder": pt(person.pose.right_shoulder),
                        "left_elbow":     pt(person.pose.left_elbow),
                        "right_elbow":    pt(person.pose.right_elbow),
                        "left_wrist":     pt(person.pose.left_wrist),
                        "right_wrist":    pt(person.pose.right_wrist),
                        "left_hip":       pt(person.pose.left_hip),
                        "right_hip":      pt(person.pose.right_hip),
                        "left_knee":      pt(person.pose.left_knee),
                        "right_knee":     pt(person.pose.right_knee),
                        "left_ankle":     pt(person.pose.left_ankle),
                        "right_ankle":    pt(person.pose.right_ankle),
                    }
                    jpose["all_landmarks"] = [pt(p) for p in person.pose.all]
                jp["pose"] = jpose

            jf["persons"].append(jp)

        data["frames"].append(jf)

    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

    print(f"[Exporter] JSON 저장 완료: {out_path}")
    return True


# ──────────────────────────────────────────────────────────────────────────
# AE Keyframe Data 내보내기
# ──────────────────────────────────────────────────────────────────────────

def _write_ae_file(path: str,
                   info: VideoInfo,
                   frames: List[FrameData],
                   getter: Callable[[FrameData], Point2D],
                   threshold: float = 0.05) -> bool:
    os.makedirs(os.path.dirname(path), exist_ok=True)

    lines = []
    lines.append("Adobe After Effects 8.0 Keyframe Data\n")
    lines.append(f"\tUnits Per Second\t{info.fps}\n")
    lines.append(f"\tSource Width\t{info.width}\n")
    lines.append(f"\tSource Height\t{info.height}\n")
    lines.append("\tSource Pixel Aspect Ratio\t1\n")
    lines.append("\tComp Pixel Aspect Ratio\t1\n")
    lines.append("\n")
    lines.append("Transform\tPosition\n")
    lines.append("\tFrame\tX pixels\tY pixels\tZ pixels\t\n")

    count = 0
    for fd in frames:
        p = getter(fd)
        if p.confidence < threshold:
            continue
        lines.append(f"\t{fd.index}\t{p.x:.4f}\t{p.y:.4f}\t0.0000\t\n")
        count += 1

    lines.append("\nEnd of Keyframe Data\n")

    with open(path, "w", encoding="utf-8") as f:
        f.writelines(lines)

    if count == 0:
        print(f"[Exporter] 경고: 유효 키프레임 없음 → {os.path.basename(path)}")
    return True


def export_ae_keyframes(frames: List[FrameData], info: VideoInfo, out_dir: str,
                        include_face: bool = True, include_body: bool = True,
                        include_hands: bool = True) -> bool:
    max_persons = max((len(fd.persons) for fd in frames), default=0)
    if max_persons == 0:
        print(f"[Exporter] 경고: 감지된 사람 없음 — AE 출력 건너뜀")
        return True

    ok = True
    face_pts = [
        ("right_eye_outer",  lambda p: p.face.right_eye_outer),
        ("right_eye_inner",  lambda p: p.face.right_eye_inner),
        ("right_iris",       lambda p: p.face.right_iris),
        ("left_eye_inner",   lambda p: p.face.left_eye_inner),
        ("left_eye_outer",   lambda p: p.face.left_eye_outer),
        ("left_iris",        lambda p: p.face.left_iris),
        ("nose_bridge_top",  lambda p: p.face.nose_bridge_top),
        ("nose_tip",         lambda p: p.face.nose_tip),
        ("mouth_right",      lambda p: p.face.mouth_right),
        ("mouth_upper",      lambda p: p.face.mouth_upper),
        ("mouth_left",       lambda p: p.face.mouth_left),
        ("mouth_lower",      lambda p: p.face.mouth_lower),
    ]
    hand_tips = [
        ("wrist",      lambda h: h.wrist),
        ("thumb_tip",  lambda h: h.thumb_tip),
        ("index_tip",  lambda h: h.index_tip),
        ("middle_tip", lambda h: h.middle_tip),
        ("ring_tip",   lambda h: h.ring_tip),
        ("pinky_tip",  lambda h: h.pinky_tip),
    ]
    pose_joints = [
        ("left_shoulder",  lambda p: p.pose.left_shoulder),
        ("right_shoulder", lambda p: p.pose.right_shoulder),
        ("left_elbow",     lambda p: p.pose.left_elbow),
        ("right_elbow",    lambda p: p.pose.right_elbow),
        ("left_wrist",     lambda p: p.pose.left_wrist),
        ("right_wrist",    lambda p: p.pose.right_wrist),
        ("left_hip",       lambda p: p.pose.left_hip),
        ("right_hip",      lambda p: p.pose.right_hip),
        ("left_knee",      lambda p: p.pose.left_knee),
        ("right_knee",     lambda p: p.pose.right_knee),
        ("left_ankle",     lambda p: p.pose.left_ankle),
        ("right_ankle",    lambda p: p.pose.right_ankle),
    ]

    for pid in range(max_persons):
        p_dir    = os.path.join(out_dir, f"person_{pid}")
        face_dir = os.path.join(p_dir, "face")
        lh_dir   = os.path.join(p_dir, "hands", "left")
        rh_dir   = os.path.join(p_dir, "hands", "right")
        lh_all   = os.path.join(lh_dir, "all")
        rh_all   = os.path.join(rh_dir, "all")
        pose_dir = os.path.join(p_dir, "pose")

        def _person(f, pid=pid):
            return f.persons[pid] if pid < len(f.persons) else None

        # ── 얼굴 12개 포인트
        if include_face:
            for name, getter in face_pts:
                ok &= _write_ae_file(
                    os.path.join(face_dir, f"{name}.txt"), info, frames,
                    lambda f, g=getter, pid=pid: (
                        g(f.persons[pid]) if pid < len(f.persons) and f.persons[pid].face.detected
                        else Point2D()))

        # ── 손 주요 6포인트 × 2손 + 전체 21포인트
        if include_hands:
            for name, hgetter in hand_tips:
                ok &= _write_ae_file(
                    os.path.join(lh_dir, f"{name}.txt"), info, frames,
                    lambda f, g=hgetter, pid=pid: (
                        g(f.persons[pid].left_hand) if pid < len(f.persons) and f.persons[pid].left_hand.detected
                        else Point2D()))
                ok &= _write_ae_file(
                    os.path.join(rh_dir, f"{name}.txt"), info, frames,
                    lambda f, g=hgetter, pid=pid: (
                        g(f.persons[pid].right_hand) if pid < len(f.persons) and f.persons[pid].right_hand.detected
                        else Point2D()))
            for i, lm_name in enumerate(HAND_LANDMARK_NAMES):
                ok &= _write_ae_file(
                    os.path.join(lh_all, f"{lm_name}.txt"), info, frames,
                    lambda f, idx=i, pid=pid: (
                        f.persons[pid].left_hand.landmarks[idx]
                        if pid < len(f.persons) and f.persons[pid].left_hand.detected
                        and len(f.persons[pid].left_hand.landmarks) > idx
                        else Point2D()))
                ok &= _write_ae_file(
                    os.path.join(rh_all, f"{lm_name}.txt"), info, frames,
                    lambda f, idx=i, pid=pid: (
                        f.persons[pid].right_hand.landmarks[idx]
                        if pid < len(f.persons) and f.persons[pid].right_hand.detected
                        and len(f.persons[pid].right_hand.landmarks) > idx
                        else Point2D()))

        # ── 포즈 주요 관절 12개
        if include_body:
            for name, getter in pose_joints:
                ok &= _write_ae_file(
                    os.path.join(pose_dir, f"{name}.txt"), info, frames,
                    lambda f, g=getter, pid=pid: (
                        g(f.persons[pid]) if pid < len(f.persons) and f.persons[pid].pose.detected
                        else Point2D()))

    print(f"[Exporter] AE Keyframe Data 저장 완료: {out_dir}/")
    return ok

File: src/Exporter.cpp
#include "Exporter.hpp"
#include <nlohmann/json.hpp>

#include <fstream>
#include <iomanip>
#include <iostream>
#include <filesystem>
#include <functional>

namespace fs = std::filesystem;
using json   = nlohmann::json;

namespace tracker {

// -----------------------------------------------------------------------
// 헬퍼: Point2D → json object
// -----------------------------------------------------------------------
static json pointToJson(const Point2D& p) {
    return {{"x", p.x}, {"y", p.y}, {"confidence", p.confidence}};
}

// -----------------------------------------------------------------------
// JSON 전체 내보내기
// -----------------------------------------------------------------------
bool Exporter::exportJSON(const std::vector<FrameData>& frames,
                          const VideoInfo&               info,
                          const std::string&             outPath) {
    fs::create_directories(fs::path(outPath).parent_path());

    json root;

    // 메타데이터
    root["metadata"] = {
        {"fps",         info.fps},
        {"total_frames", info.totalFrames},
        {"width",       info.width},
        {"height",      info.height}
    };

    // 랜드마크 이름 테이블 (After Effects 참조용)
    root["landmark_info"] = {
        {"face", {
            {"total_points", 70},
            {"named", {
                {"right_eye_outer_corner", 36},
                {"right_eye_inner_corner", 39},
                {"right_pupil",            68},
                {"left_eye_inner_corner",  42},
                {"left_eye_outer_corner",  45},
                {"left_pupil",             69},
                {"nose_bridge_top",        27},
                {"nose_tip",               33},
                {"mouth_right_corner",     48},
                {"mouth_upper_center",     51},
                {"mouth_left_corner",      54},
                {"mouth_lower_center",     57}
            }}
        }},
        {"hand", {
            {"total_points", 21},
            {"names", {
                "wrist",
                "thumb_cmc","thumb_mcp","thumb_ip","thumb_tip",
                "index_mcp","index_pip","index_dip","index_tip",
                "middle_mcp","middle_pip","middle_dip","middle_tip",
                "ring_mcp","ring_pip","ring_dip","ring_tip",
                "pinky_mcp","pinky_pip","pinky_dip","pinky_tip"
            }}
        }}
    };

    // 프레임별 데이터
    json framesArr = json::array();
    for (const auto& fd : frames) {
        json jf;
        jf["frame"]     = fd.index;
        jf["timestamp"] = fd.timestamp;

        // --- 얼굴 ---
        json jface;
        jface["detected"] = fd.face.detected;
        if (fd.face.detected) {
            jface["named"] = {
                {"right_eye_outer_corner", pointToJson(fd.face.rightEyeOuterCorner)},
                {"right_eye_inner_corner", pointToJson(fd.face.rightEyeInnerCorner)},
                {"right_pupil",            pointToJson(fd.face.rightPupil)},
                {"left_eye_inner_corner",  pointToJson(fd.face.leftEyeInnerCorner)},
                {"left_eye_outer_corner",  pointToJson(fd.face.leftEyeOuterCorner)},
                {"left_pupil",             pointToJson(fd.face.leftPupil)},
                {"nose_bridge_top",        pointToJson(fd.face.noseBridgeTop)},
                {"nose_tip",               pointToJson(fd.face.noseTip)},
                {"mouth_right_corner",     pointToJson(fd.face.mouthRightCorner)},
                {"mouth_upper_center",     pointToJson(fd.face.mouthUpperCenter)},
                {"mouth_left_corner",      pointToJson(fd.face.mouthLeftCorner)},
                {"mouth_lower_center",     pointToJson(fd.face.mouthLowerCenter)}
            };

            // 전체 70개 raw
            json allArr = json::array();
            for (const auto& p : fd.face.all) allArr.push_back(pointToJson(p));
            jface["all_landmarks"] = allArr;
        }
        jf["face"] = jface;

        // --- 손 ---
        auto handToJson = [](const HandLandmarks& h) {
            json jh;
            jh["detected"] = h.detected;
            jh["side"]     = h.side;
            if (h.detected) {
                static const char* names[] = {
                    "wrist",
                    "thumb_cmc","thumb_mcp","thumb_ip","thumb_tip",
                    "index_mcp","index_pip","index_dip","index_tip",
                    "middle_mcp","middle_pip","middle_dip","middle_tip",
                    "ring_mcp","ring_pip","ring_dip","ring_tip",
                    "pinky_mcp","pinky_pip","pinky_dip","pinky_tip"
                };
                json lmArr = json::array();
                for (int i = 0; i < (int)h.landmarks.size(); ++i) {
                    json lm = pointToJson(h.landmarks[i]);
                    lm["index"] = i;
                    if (i < 21) lm["name"] = names[i];
                    lmArr.push_back(lm);
                }
                jh["landmarks"] = lmArr;
            }
            return jh;
        };

        jf["left_hand"]  = handToJson(fd.leftHand);
        jf["right_hand"] = handToJson(fd.rightHand);

        framesArr.push_back(jf);
    }
    root["frames"] = framesArr;

    // 파일 저장
    std::ofstream ofs(outPath);
    if (!ofs.is_open()) {
        std::cerr << "[Exporter] JSON 파일 쓰기 실패: " << outPath << "\n";
        return false;
    }
    ofs << std::setw(2) << root << "\n";
    std::cout << "[Exporter] JSON 저장 완료: " << outPath << "\n";
    return true;
}

// -----------------------------------------------------------------------
// AE Keyframe Data 단일 파일 작성
// -----------------------------------------------------------------------
bool Exporter::writeAEFile(const std::string&              path,
                           const std::string&              layerName,
                           const VideoInfo&                info,
                           const std::vector<FrameData>&   frames,
                           std::function<Point2D(const FrameData&)> getter) {
    fs::create_directories(fs::path(path).parent_path());

    std::ofstream ofs(path);
    if (!ofs.is_open()) {
        std::cerr << "[Exporter] AE 파일 쓰기 실패: " << path << "\n";
        return false;
    }

    ofs << std::fixed << std::setprecision(4);

    // ── 헤더 ──────────────────────────────────────────────────────────
    ofs << "Adobe After Effects 8.0 Keyframe Data\n\n";
    ofs << "\tUnits Per Second\t" << info.fps   << "\n";
    ofs << "\tSource Width\t"     << info.width  << "\n";
    ofs << "\tSource Height\t"    << info.height << "\n";
    ofs << "\tSource Pixel Aspect Ratio\t1\n";
    ofs << "\tComp Pixel Aspect Ratio\t1\n\n";

    // ── 키프레임 섹션 ─────────────────────────────────────────────────
    ofs << "Transform\tPosition\n";
    ofs << "\tFrame\tX pixels\tY pixels\tZ pixels\t\n";

    int written = 0;
    for (const auto& fd : frames) {
        Point2D p = getter(fd);
        if (!p.valid(0.05f)) continue;   // 미검출 프레임 스킵 (AE가 보간)

        ofs << "\t" << fd.index
            << "\t" << p.x
            << "\t" << p.y
            << "\t" << "0.0000" << "\t\n";
        ++written;
    }

    ofs << "\nEnd of Keyframe Data\n";

    if (written == 0) {
        std::cout << "[Exporter] 경고: '" << layerName << "' 는 유효한 키프레임 없음\n";
    }
    return true;
}

// -----------------------------------------------------------------------
// AE Keyframe Data 전체 내보내기
// -----------------------------------------------------------------------
bool Exporter::exportAEKeyframes(const std::vector<FrameData>& frames,
                                 const VideoInfo&               info,
                                 const std::string&             outDir) {
    const std::string faceDir      = outDir + "/face";
    const std::string handLeftDir  = outDir + "/hands/left";
    const std::string handRightDir = outDir + "/hands/right";

    bool ok = true;

    // ── 얼굴 (12 포인트) ─────────────────────────────────────────────
    struct FaceEntry {
        std::string name;
        std::function<Point2D(const FrameData&)> getter;
    };

    const std::vector<FaceEntry> facePoints = {
        {"right_eye_outer_corner", [](const FrameData& f){ return f.face.rightEyeOuterCorner; }},
        {"right_eye_inner_corner", [](const FrameData& f){ return f.face.rightEyeInnerCorner; }},
        {"right_pupil",            [](const FrameData& f){ return f.face.rightPupil;           }},
        {"left_eye_inner_corner",  [](const FrameData& f){ return f.face.leftEyeInnerCorner;  }},
        {"left_eye_outer_corner",  [](const FrameData& f){ return f.face.leftEyeOuterCorner;  }},
        {"left_pupil",             [](const FrameData& f){ return f.face.leftPupil;            }},
        {"nose_bridge_top",        [](const FrameData& f){ return f.face.noseBridgeTop;        }},
        {"nose_tip",               [](const FrameData& f){ return f.face.noseTip;              }},
        {"mouth_right_corner",     [](const FrameData& f){ return f.face.mouthRightCorner;     }},
        {"mouth_upper_center",     [](const FrameData& f){ return f.face.mouthUpperCenter;     }},
        {"mouth_left_corner",      [](const FrameData& f){ return f.face.mouthLeftCorner;      }},
        {"mouth_lower_center",     [](const FrameData& f){ return f.face.mouthLowerCenter;     }},
    };

    for (const auto& fp : facePoints) {
        ok &= writeAEFile(faceDir + "/" + fp.name + ".txt",
                          fp.name, info, frames, fp.getter);
    }

    // ── 손 — 주요 6개 포인트 × 2 (왼/오른) ──────────────────────────
    struct HandEntry {
        std::string name;
        std::function<Point2D(const HandLandmarks&)> getter;
        int lmIdx; // for all-landmarks export
    };

    const std::vector<HandEntry> handPoints = {
        {"wrist",      [](const HandLandmarks& h){ return h.wrist();      }, 0},
        {"thumb_tip",  [](const HandLandmarks& h){ return h.thumbTip();   }, 4},
        {"index_tip",  [](const HandLandmarks& h){ return h.indexTip();   }, 8},
        {"middle_tip", [](const HandLandmarks& h){ return h.middleTip();  }, 12},
        {"ring_tip",   [](const HandLandmarks& h){ return h.ringTip();    }, 16},
        {"pinky_tip",  [](const HandLandmarks& h){ return h.pinkyTip();   }, 20},
    };

    for (const auto& hp : handPoints) {
        // 왼손
        ok &= writeAEFile(handLeftDir  + "/" + hp.name + ".txt",
                          "left_"  + hp.name, info, frames,
                          [&hp](const FrameData& f){ return hp.getter(f.leftHand);  });
        // 오른손
        ok &= writeAEFile(handRightDir + "/" + hp.name + ".txt",
                          "right_" + hp.name, info, frames,
                          [&hp](const FrameData& f){ return hp.getter(f.rightHand); });
    }

    // ── 전체 손 랜드마크 21개 (선택적 — 상세 제어용) ─────────────────
    static const char* handLMNames[] = {
        "wrist",
        "thumb_cmc","thumb_mcp","thumb_ip","thumb_tip",
        "index_mcp","index_pip","index_dip","index_tip",
        "middle_mcp","middle_pip","middle_dip","middle_tip",
        "ring_mcp","ring_pip","ring_dip","ring_tip",
        "pinky_mcp","pinky_pip","pinky_dip","pinky_tip"
    };

    for (int i = 0; i < 21; ++i) {
        const int idx = i;
        ok &= writeAEFile(handLeftDir  + "/all/" + std::string(handLMNames[i]) + ".txt",
                          "left_"  + std::string(handLMNames[i]), info, frames,
                          [idx](const FrameData& f) -> Point2D {
                              if (!f.leftHand.detected || (int)f.leftHand.landmarks.size() <= idx)
                                  return {};
                              return f.leftHand.landmarks[idx];
                          });
        ok &= writeAEFile(handRightDir + "/all/" + std::string(handLMNames[i]) + ".txt",
                          "right_" + std::string(handLMNames[i]), info, frames,
                          [idx](const FrameData& f) -> Point2D {
                              if (!f.rightHand.detected || (int)f.rightHand.landmarks.size() <= idx)
                                  return {};
                              return f.rightHand.landmarks[idx];
                          });
    }

    std::cout << "[Exporter] AE Keyframe Data 저장 완료: " << outDir << "\n";
    return ok;
}

} // namespace tracker
